pkg/metrics/filteredlabels.go

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package metrics
+
+import (
+	"github.com/cilium/tetragon/pkg/option"
+)
+
+type FilteredLabels interface {
+	Keys() []string
+	Values() []string
+}
+
+type ProcessLabels struct {
+	Namespace string
+	Workload  string
+	Pod       string
+	Binary    string
+}
+
+func NewProcessLabels(namespace, workload, pod, binary string) *ProcessLabels {
+	if !option.Config.MetricsLabelFilter["namespace"] {
+		namespace = ""
+	}
+	if !option.Config.MetricsLabelFilter["workload"] {
+		workload = ""
+	}
+	if !option.Config.MetricsLabelFilter["pod"] {
+		pod = ""
+	}
+	if !option.Config.MetricsLabelFilter["binary"] {
+		binary = ""
+	}
```

Comment:
> 1. Configure MetricsLabelFilter similarly like before, but as map[string]bool
> 2. metrics.ProcessLabels struct implements metrics.FilteredLabels interface and
>    contains label values of configurable labels. Disabled labels are set to ""
>    when the struct is instantiated.
> 3. Define a metric using a "granular metric" wrapper, which now uses Go generics
>    to specify configurable labels (FilteredMetrics type).
> 4. The inner Prometheus metric is now registered with all labels.
> 5. Then we update the metric using a slightly different implementation of
>    WithLabelValues, which takes a generic FileredLabels struct and any
>    additional label values as strings.

AFAIU, the disabled labels are set to "". The thing I'm not sure I understand is where does the filtering happens? Is it .WithLabelValues()?

Reply:
When calling `WithLabelValues()` you need to pass a `ProcessLabels` struct. If it's created with `NewProcessLabels` then disabled labels are set to "", what from Prometheus perspective is equivalent to not having these labels at all. So filtering relies on always creating `ProcessLabels` with `NewProcessLabels`. I'll add a comment about it, as it's not obvious indeed.